csrc/device_lower/pass/expr_sort.cpp

```diff
@@ -889,13 +997,20 @@ ExprGroup* ExprSegmentationSorter::makeMergedNode(
   return joined_groups;
 }
 
-bool canReducePA(ExprGroup* group) {
+bool ExprSegmentationSorter::canReducePA(ExprGroup* group) const {
   if (group->payload()->pa_domains.empty()) {
     return false;
   }
 
   IterDomain* group_pa_last_id = group->payload()->pa_domains.back();
 
+  // If the last ID is the kernel scope, there should still be
+  // consumer groups that use tensors produced by this group, so we
+  // should not remove it from the PA set yet
```

Comment:
You're right. I was confused with CA and PA. Will fix the comment.

In the case of `T1[0] = set(scalar);`, this expr won't have the kernel scope in its PA domain vector as the input is a scalar.

https://github.com/NVIDIA/Fuser/pull/614/files/d123cfbd428e661a411f9f2544a3bca48286a623#diff-08d42baef83b99f8e762caa216f02bf1740b12cf554756242f24de1b0e940cccR952

I did this since I think we don't need the hierarchical ordering constraint with scalars.

Reply:
Am I understanding correct that, if we have
```
scalar1 = set(scalar0);
T1[] = add(T0[], scalar1);
```
we will have two groups `{scalar1 = set(scalar0)}` and `{T1[] = add(T0[], scalar1)}`, and we will have to reduce the PA of the second group to `0` in order to be merged with the first group?

Reply:
If `T0` is a fusion input, the group of `{T1[] = add(T0[], scalar1)}` won't have the kernel scope as part of its PA domain vector, so there's nothing to reduce. 

If `T0` is not a fusion input, we will add the kernel scope to the PA vector of `T1` and the CA vector of `T0`, and they will be first merged before `{scalar1 = set(scalar0)}`.

Reply:
What if `T0` is from a tensor factory?

Reply:
Then `T0` is not a fusion input, so the PA vector of `{T1[] = add(T0[], scalar1)}` will have the kernel scope as well as the CA vector of the `T0` factory expression.

Reply:
OK, then `{T1[] = add(T0[], scalar1)}` will be merged with `{T0[] = full(scalar2)}` first, and the resulting group will have empty PA, which can then be merged with `{scalar1 = set(scalar0)}`?

Reply:
Exactly